apps/serdes/Serializer.cpp

```diff
@@ -0,0 +1,755 @@
+    }
+    case Halide::Internal::IRNodeType::AssertStmt: {
+        auto assert_stmt = stmt.as<Halide::Internal::AssertStmt>();
+        auto condition = assert_stmt->condition;
+        auto condition_serialized = serialize_expr(builder, condition);
+        auto message = assert_stmt->message;
+        auto message_serialized = serialize_expr(builder, message);
+        return std::make_pair(Halide::Serdes::Stmt::Stmt_AssertStmt, Halide::Serdes::CreateAssertStmt(builder, condition_serialized.first, condition_serialized.second, message_serialized.first, message_serialized.second).Union());
+    }
+    case Halide::Internal::IRNodeType::ProducerConsumer: {
+        auto producer_consumer = stmt.as<Halide::Internal::ProducerConsumer>();
+        std::string name = producer_consumer->name;
+        auto name_serialized = serialize_string(builder, name);
+        bool is_producer = producer_consumer->is_producer;
+        auto body = producer_consumer->body;
+        auto body_serialized = serialize_stmt(builder, body);
+        return std::make_pair(Halide::Serdes::Stmt::Stmt_ProducerConsumer, Halide::Serdes::CreateProducerConsumer(builder, name_serialized, is_producer, body_serialized.first, body_serialized.second).Union());
+    }
+    case Halide::Internal::IRNodeType::For: {
+        auto for_stmt = stmt.as<Halide::Internal::For>();
+        std::string name = for_stmt->name;
+        auto name_serialized = serialize_string(builder, name);
+        auto min = for_stmt->min;
+        auto min_serialized = serialize_expr(builder, min);
+        auto extent = for_stmt->extent;
+        auto extent_serialized = serialize_expr(builder, extent);
+        Halide::Serdes::ForType for_type = Halide::Serdes::ForType::ForType_Serial;
+        switch (for_stmt->for_type) {
+        case Halide::Internal::ForType::Serial: {
+            for_type = Halide::Serdes::ForType::ForType_Serial;
+            break;
+        }
+        case Halide::Internal::ForType::Parallel: {
+            for_type = Halide::Serdes::ForType::ForType_Parallel;
+            break;
+        }
+        case Halide::Internal::ForType::Vectorized: {
+            for_type = Halide::Serdes::ForType::ForType_Vectorized;
+            break;
+        }
+        case Halide::Internal::ForType::Unrolled: {
+            for_type = Halide::Serdes::ForType::ForType_Unrolled;
+            break;
+        }
+        case Halide::Internal::ForType::Extern: {
+            for_type = Halide::Serdes::ForType::ForType_Extern;
+            break;
+        }
+        case Halide::Internal::ForType::GPUBlock: {
+            for_type = Halide::Serdes::ForType::ForType_GPUBlock;
+            break;
+        }
+        case Halide::Internal::ForType::GPUThread: {
+            for_type = Halide::Serdes::ForType::ForType_GPUThread;
+            break;
+        }
+        case Halide::Internal::ForType::GPULane: {
+            for_type = Halide::Serdes::ForType::ForType_GPULane;
+            break;
+        }
```

Comment:
Here and elsewhere: how do we ensure that additions to one of the enums (in either the `.fbs` or the Halide headers files) are correctly handled? I don't know the answer to this (and I'm not expecting you to have the answer, either), but I hope there exists an answer that is more robust than "hope we don't forget someday".

Reply:
I think we can alleviate this by removing the default case of the switch statement, and do the error handling outside of the switch. Then whenever we have a new addition to the enums in Halide headers (I think the workflow will be starting with adding to Halide files and then serialization protocol `fbs` file), the compiler will give an error here because of non-exhaustive enumeration